fix issues when using module in a browser.

mfrisbey opened the pull request:
## Description

Provides various changes to allow this module to be used in both a browser and Node.JS context:

* Refactor usage of `posix` from the `path` module, since it's not available when run in a browser.
* Support passing a `blob` option when uploading files with `AEMUpload`.
* Fix cases where the module was assuming the response body of a request would always be a stream; it's not a stream when running in a browser.
* Changed so that `window.fetch` is used instead of `node-fetch-npm` when running in a browser. `node-fetch-npm` was only designed to run in Node.JS, so it wasn't working correctly in this case.

Fixes #69 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- ~[ ] My change requires a change to the documentation.~
- ~[ ] I have updated the documentation accordingly.~
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mfrisbey commented:
Note that I'm running some tests on Windows to make sure that the path handling still works as expected. Will update the PR if additional changes needed.

tmathern commented on lib/aem/aemupload.js:
```diff
@@ -33,6 +33,7 @@ const { RandomFileAccess } = require("../randomfileaccess");
 const EventEmitter = require("events");
 const UploadError = require("./upload-error");
 const { FilterFailedAssets } = require("../functions/filterfailedassets");
+const { IllegalArgumentError } = require('../error');
```
Nitpick: to be consistent with the other imports, single quotes should be turned into double-quotes.
```suggestion
const { IllegalArgumentError } = require("../error");
```

mfrisbey replied:
Good catch on the quotes, I'll fix those.

tmathern commented on lib/aem/aemupload.js:
```diff
@@ -43,7 +44,16 @@ const { FilterFailedAssets } = require("../functions/filterfailedassets");
+            if (!uploadFile.filePath) {
+                throw new IllegalArgumentError(
+                    'Either blob or filePath must be provided in uploadFiles',
+                    JSON.stringify(uploadFile),
```
eslint may complain about this if the last comma is not removed:
```suggestion
                    JSON.stringify(uploadFile)
```

mfrisbey replied:
It didn't complain, but I'll remove the comma anyway.

tmathern commented on lib/fetch.js:
```diff
@@ -138,6 +147,40 @@ async function stream(method, url, options) {
+ */
+async function handleStream(method, url, options) {
+    const response = await stream(method, url, options);
+    if (response.body && response.body.on) {
```
Can response be undefined/null?
```suggestion
    if (response && response.body && response.body.on) {
```

mfrisbey replied:
I think we'd have bigger problems downstream if it were, but I'll add the check anyway to be safe.

tmathern reviewed (approved):
Some smaller changes - but otherwise LGTM.

Where were the errors observed?

tmathern commented:
Do all the browsers supported have fetch now? Even the older ones?

mfrisbey commented:
> @tmathern Do all the browsers supported have fetch now? Even the older ones?

It appears that the modern browsers [support it](https://caniuse.com/fetch). There are some edge cases that don't, but IMO what we have should be sufficient for the vast majority of cases.

After some testing on windows, I opted to [remove posix entirely](https://github.com/adobe/node-httptransfer/pull/70/files#diff-3fb720f70ea34fb2975fd6c38ac4b6b6131373be2abfa4760dfaae8a25daaee2R175), and just replace all occurrences of `\` with `/` when needing a URL path. Admittedly, I don't understand the original motivation for using `posix`, so I wanted to get a second opinion on that particular change. My testing showed that this strategy was more robust on Windows, making it so that either `/` or `\` will work. WDYT?

bobvanmanen commented on lib/asset/asset.js:
```diff
@@ -13,7 +13,8 @@
 
 require("core-js/stable");
-const { dirname: urlPathDirname, basename: urlPathBasename } = require("path").posix;
+const { basename: filePathBasename } = require("path");
```
@mfrisbey I picked `posix` with the thought that it would behave the same whether it runs on Windows or Mac. Technically on Windows the path separator is `\` and not `/` which is the separator in urls. It looks like you already verified this though.

bobvanmanen commented:
Just to bring it up, but how do we ensure going forward that this keeps working on Node.js Mac, Node.js Windows, and in modern browsers? It seems easy to break unless we keep manually testing it which doesn't scale.

tmathern commented:
@bobvanmanen @mfrisbey 
More on integration than unit tests side:
There is a headless scriptable browser, [pahntomJS](https://phantomjs.org/) that can be used to do simple testing. Otherwise, there is Selenium to automate Chrome/Safari/Edge (Chrome has a headless version too). But I don't think any UI validation is needed here, just the workings without UI-related updates.

mfrisbey commented:
> @bobvanmanen how do we ensure going forward that this keeps working on Node.js Mac, Node.js Windows, and in modern browsers?

Here are some of my thoughts:

* Ideally, it would be awesome if the repo's CI could run a set of e2e tests on different images (i.e. Mac, Windows, and a browser). Unfortunately, the complexity of setting up an instance of AEM to test against may prevent this unless we somehow set up a dedicated cloud environment that the tests can run against.
* A slightly less optimal solution could be to set up some tests (similar to [aem-upload](https://github.com/adobe/aem-upload/tree/master/e2e)) that are run locally before creating PRs. A single set of node.js tests could be run on Mac/Windows, and we could create some kind of simple React application that could be run to test basic browser support.
* As a bare minimum, I added more unit tests that cover the issues that I fixed in this PR, so there is at least better unit test coverage for the browser case now.
